Gossipsub: Partial Message Extension (#685)
libp2p/specs: pubsub/gossipsub/partial-messages.md

application generates this and consumes this.

A later `PartialIWANT` serve to refine the request of prior a prior `PartialIWANT`.

Nodes SHOULD assume a `PartialIWANT` implies a `IDONTWANT` for the full message.

### PartialIHAVE

A `PartialIHAVE` allows nodes to signal HAVE information before receiving all
segments, unlocking the use of `PartialIWANT` in more contexts.

In the context of partial messages, it is more useful than IHAVE as it includes
the group ID. In contrast, an IHAVE does not. A receiving peer has no way to
link an IHAVE's message ID with a group ID, without having the full message.

A `PartialIHAVE` message can be used both in the context of lazy push, notifying
peers about available parts, and in the context of heartbeats as a replacement
to IHAVEs.

The structure of `PartialIHAVE` is analogous to that of `PartialIWANT`.

The metadata, as in a `PartialIWANT`, is application defined. It is some encoding
that represents the parts the sender has.

Implementations are free to select when to send an update to their peers based
on signaling bandwidth tradeoff considerations.

Receivers MUST treat a `PartialIHAVE` as a signal that the peer does not want

> the indicated part.

Author: I think this is covered now.

### Changes to `SubOpts` and interaction with the existing Gossipsub mesh.

Partial Messages uses the same mesh as normal Gossipsub messages. It is a
replacement to "full" messages. A node requests a peer to use partial messages
for a specific topic by setting the `partial` field in the `SubOpts` message.
The `SubOpts` message is how a peer subscribes to a topic.

If a node receives a subscribe request with the `partial` field set to true, it
MUST send partial messages instead of full messages.

It is an error to set the partial field true if the peer does not support
partial extensions.

The partial field value MUST be ignored when a peer sends an unsubscribe message
`SubOpts.subscribe=false`.

## Application Interface

Message contents are application defined. Thus splitting a message must be
application defined. Applications should provide a Partial Message type that

> supports the following operations:

Reviewer: I'd move the non normative disclaimer here.
 
For the rest of the text, I see two implementation strategies: 
1. Discard all gossipsub infra except sending and receiving messages, and topic mesh handling, and do all the partial message stuff in the application. So all you need from the application is `PublishTo(msg, to)` and `ReceiveRPC(from, rpc)`

2. The one that go-libp2p does where the application periodically does a PublishMessage(msg) providing its view of a particular partial message(topicID, groupID) to the gossipsub layer. The gs layer keeps track of the want and have lists per peer as opaque metadata and updates it based on the methods provided below. 

You may rephrase the text above or use something similar. When reading the document I assumed we'd do 1, and so found this section confusing, but reviewing https://github.com/libp2p/go-libp2p-pubsub/pull/631/files#diff-d45a19dbc01044648070c29c0fc7b2b63edd3deaccf85b797314d5ef991f4db6R48-R51 I can see we do 2, which clears up doubts on my initial readings, like https://github.com/libp2p/specs/pull/685/files#r2379080632

Author: We mostly do 2. We try to do as much as we can in the gossipsub layer without making too many assumptions.

1. `.GroupID() -> GroupID: bytes`

> 2. `.PartialMessageBytesFromMetadata(metadata: bytes) -> Result<(EncodedPartialMessage: bytes, metadata: bytes), Error>` (When responding to a `PartialIWANT` or eagerly pushing a partial message)

Reviewer: NIT: I prefer `repr` over `metadata` as it is the IWant byte representation, not some meta info about it.

> 2. `.PartialMessageBytesFromMetadata(metadata: bytes) -> Result<(EncodedPartialMessage: bytes, metadata: bytes), Error>` (When responding to a `PartialIWANT` or eagerly pushing a partial message)

Reviewer: NIT: 
```suggestion
2. `.PartialMessageBytesFromMetadata(IWantMetadata: bytes) -> Result<(EncodedPartialMessage: bytes, newIWantMetadata: bytes), Error>` (When responding to a `PartialIWANT` or eagerly pushing a partial message)
```

  a. The returned metadata represents the still missing parts. For example, if a
     peer is only able to fulfill a part of the the request, the returned
     metadata represents the parts it couldn't fulfill.
3. `.ExtendFromEncodedPartialMessage(data: bytes) -> Result<(), Error>` (When receiving a `PartialMessage`)
4. `.MissingParts() -> Result<metadata: bytes, Error>` (For `PartialIWANT`)
5. `.AvailableParts() -> Result<metadata: bytes, Error>` (For `PartialIHAVE`)

Gossipsub in turn provides a `.PublishPartial(PartialMessage)` method.

Note that this specific interface is not intended to be normative to
implementations, rather, it is high level summary of what each layer should
provide.

## Protobuf

Refer to the protobuf registry at `./extensions/extensions.proto`

## Open Questions

- Do we want to add a TTL to PartialIWANTs? This would allow us to cancel them after some time.
- Should we rename the metadata bytes to iwant and ihave?

> - In the bitmap usecase, iwant/ihave are simply inverses of each other. Do we need to send them both?

Author: Done in 8e42e179.

I think this simplifies the implementation quite a bit as well.

and spec status.

[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md

## Overview

Partial Messages Extensions allow users to transmit only a small part of a
message rather than a full message. This is especially useful in cases where
this a large messages and a peer is missing only a small part of the message.

Much of the complexity around partial messages is contained by the Gossipsub
implementation. Applications require little changes to benefit from this
extension.

## Terms and Definitions

**Full Message**: A Gossipsub Message.

**Partial Message**: Some part of a Full Message. Each part should be able to be
validated.

**Group ID**: An identifier to some Full Message. This must not depend on
knowing the full message, so it can not simply be a hash of the full message.

**Group of Partial Messages**: A set of partial messages all belonging to the
same group. The group may be complete, meaning that all parts are available. Or
it may be incomplete, meaning some parts are missing.

## Motivation

The main motivation for this extension is optimizing Ethereum's Data

> Availability Sampling (DAS) protocol. In Ethereum's upcoming fork, Fusaka,

Reviewer 2: Technically this doesn't benefit the Sampling part of things. Maybe just generalize this to Ethereum's DA protocols.

> custodied data is laid out in a matrix, where the rows represent user data

Reviewer 2: ```suggestion
custodied data is laid out in a matrix per block, where the rows represent user data
```

> (called blobs), and the columns represent a slice across all blobs (each blob

Reviewer 2: ```suggestion
(called blobs), and the columns represent a slice across all blobs included in the block (each blob
```

64KiB per column. and since nodes custody at least 8 columns, the total savings
per slot is around 500KiB, or 4 Megabits per slot.

Later, partial messages could enable further optimizations:
- If cells can be validated individually, as in the case of DAS, partial messages
could also be forwarded, allowing us to reduce the store-and-forward delay [2].
- Finally, in the FullDAS construct, where both row and column topics are
defined, partial messages allow cross-forwarding cells between these topics [2].

## Advantage of Partial Messages over smaller Gossipsub Messages

Partial Messages within a group imply some structure and correlation. Thus,
multiple partial messages can be referenced succinctly. For example, parts can
be referenced by bitmaps, ranges, or a bloom filter.

The structure of partial messages in a group, as well as how partial messages
are referenced is application defined.

If, in some application, a group only ever contained a single partial message,
then partial messages would be the same as smaller messages.

## Protocol Messages

The following section specifies the semantics of each new protocol message.

### PartialIWANT

> Partial IWants signal to a receiver that the sending peer only wants a part of

Reviewer 2: ```suggestion
`PartialIWANT`s signal to a receiver that the sending peer only wants a part of
```

some message.

The message to which the peer is requesting a part of is identified by the
`groupID` identifier. This is similar to a complete message's `MessageID`, but,
in contrast to a content-based message id, does not require the full message to
compute. For example, in the Ethereum use case, this could simply be the hash of
the signed block header.

The `topicID` references the Gossipsub topic a message, and thus its parts,
belong to.

The `metadata` field is opaque application defined metadata associated with this
request. This can be a bitmap, a list of ranges, or a bloom filter. The
application generates this and consumes this.

> Nodes SHOULD assume a `partialIWANT `implies a `IDONTWANT `for the full message.

Reviewer 2: ```suggestion
Nodes SHOULD assume a `partialIWANT` implies a `IDONTWANT `for the full message.
```

### PartialIDONTWANT

PartialIDONTWANT serves to cancel any pending PartialIWANTs.

A PartialIDONTWANT without metadata serves to cancel all pending PartialIWANTs,
and signals there is no part this peer wants.

Implementations SHOULD NOT send a PartialIHAVE to a peer with parts that the
peer has previously sent a PartialIDONTWANT for.

### PartialIHAVE

PartialIHave allows nodes to signal HAVE information before receiving all
segments, unlocking the use of partialIWANT in more contexts.

In the context of partial messages, it is more useful than IHAVE as it tells the

> peer the group ID. In contrast, a IHAVE only includes a message ID that is

Reviewer 2: ```suggestion
peer the group ID. In contrast, an IHAVE only includes a message ID that is
```

unique to the message. A receiving peer has no way to link an IHAVE's message ID
with a group ID, without having the full message.

Partial IHAVE messages can be used both in the context of lazy push, notifying
peers about reception progress, and in the context of heartbeats, sending
also Partial IHAVEs.

The structure of PartialIHAVE is analogous to that of PartialIWANT.

The metadata, as in Partial IWants, is application defined. It is some encoding
that represents the parts the sender has.

Implementations are free to select when to send an update to their peers based
on signaling bandwidth tradeoff considerations.

## Application Interface

Message contents are application defined. Thus splitting a message must be
application defined. Here are a list of operations an application is expected to
provide to Gossipsub to enable partial message delivery.

> 1. Split a full message into a group of partial messages.

Reviewer 2: We may not have a **full** message, but still benefit from splitting what we have into pieces that can be trasmitted as parts right? (e.g. the common case for PeerDAS columns)

Also, is the term "partial message" is being used ambiguously in this spec? Here it refers to parts, but below it might refer to a collection of parts in response to a PARTIALIWANT?

> 2. Given a group of partial messages, encode a request for the missing parts. (Used by `PartialIWANT`)

Reviewer 2: This will require some higher level contextual information also, right? In our case, in addition to the actual partial messages generated in (1), we need to know the total number of parts and the positions our local parts occupy within the series, so we can generate the bitmap.

3. Given the request above and a group of partial messages, return the relevant
   parts of the message. (Used to fulfill a `PartialIWANT`)
4. Given a group of partial messages, encode metadata for the parts present. (Used by `PartialIHAVE`)
5. Validate a partial message.
6. Encode and decode a group of partial messages.

## Protobuf

```protobuf
syntax = "proto2";

message PartialMessagesExtension {
  optional PartialMessage message = 1;
  optional PartialIWANT iwant = 2;
  optional PartialIDONTWANT idontwant = 3;
  optional PartialIHAVE ihave = 4;
}

message PartialMessage {
  optional bytes topicID = 1;
  optional bytes data = 2;

> }

Reviewer 2: Would the partial message benefit some contextual metadata to indicate which parts are actually being delivered? Or do we expect the data to encode this (e.g. with null elements inside a list, if that's the shape)? Or does the `PartialMessage` refer to a single part?

Also: this schema only allows for one inflight partial request per peer per topic. I guess that's not a problem and we (Ethereum) are fine with that limitation?

# Partial Messages Extension

| Lifecycle Stage | Maturity      | Status | Latest Revision |
| --------------- | ------------- | ------ | --------------- |
| 1A              | Working Draft | Active | r0, 2025-06-23  |

Authors: [@marcopolo, @cskiraly]

Interest Group: TODO

[@marcopolo]: https://github.com/marcopolo
[@cskiraly]: https://github.com/cskiraly

See the [lifecycle document][lifecycle-spec] for context about the maturity level
and spec status.

[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md

## Overview

Partial Messages Extensions allow users to transmit only a small part of a
message rather than a full message. This is especially useful in cases where

> there is a large messages and a peer is missing only a small part of the

Reviewer 3: ```suggestion
there is a large message and a peer is missing only a small part of the
```

**Group ID**: An identifier to some Full Message. This must not depend on
knowing the full message, so it can not simply be a hash of the full message.

## Motivation

The main motivation for this extension is optimizing Ethereum's Data
Availability (DA) protocol. In Ethereum's upcoming fork, Fusaka, custodied data
is laid out in a matrix per block, where the rows represent user data (called
blobs), and the columns represent a slice across all blobs included in the block
(each blob slice in the column is called a cell). These columns are propagated
with Gossipsub. At the time of writing it is common for a node to already have
all the blobs from its mempool, but in cases where it doesn't (~38%[1]) have
_all_ of the blobs it almost always has _most_ of the blobs (today, it almost
always has all but one [1]).

This extension would allow nodes to only request the column message part
belonging to the missing blob. Reducing the network resource usage
significantly. As an example, if there are 32 blob cells in a column and the
node has all but one cell, this would result in a transfer of 2KiB rather than
64KiB per column. and since nodes custody at least 8 columns, the total savings

> per slot is around 500KiB, or 4 Megabits per slot.

Author: 500KiB*8=4mb. Bits since we are talking about bandwidth, but we can just remove that

link an IHAVE's message ID with a group ID, without having the full message.

A `PartialIHAVE` message can be used both in the context of lazy push, notifying
peers about available parts, and in the context of heartbeats as a replacement
to IHAVEs.

The structure of `PartialIHAVE` is analogous to that of `PartialIWANT`.

The metadata, as in a `PartialIWANT`, is application defined. It is some encoding
that represents the parts the sender has.

Implementations are free to select when to send an update to their peers based
on signaling bandwidth tradeoff considerations.

## Application Interface

Message contents are application defined. Thus splitting a message must be
application defined. Applications should provide a Partial Message type that
supports the following operations:

1. `.GroupID() -> GroupID: bytes`
2. `.PartialMessageBytesFromMetadata(metadata: bytes) -> Result<(EncodedPartialMessage: bytes, metadata: bytes), Error>` (When responding to a `PartialIWANT` or eagerly pushing a partial message)
  a. The returned metadata represents the still missing parts. For example, if a
     peer is only able to fulfill a part of the the request, the returned

>      metadata represents the parts it couldn't fulfill.

Author: Suggestion for text to make this clearer?

>      metadata represents the parts it couldn't fulfill.

Author: I rephrased this, is it clearer?

3. `.ExtendFromEncodedPartialMessage(data: bytes) -> Result<(), Error>` (When receiving a `PartialMessage`)
4. `.MissingParts() -> Result<metadata: bytes, Error>` (For `PartialIWANT`)
5. `.AvailableParts() -> Result<metadata: bytes, Error>` (For `PartialIHAVE`)

Gossipsub in turn provides a `.PublishPartial(PartialMessage)` method.

> 

Author: the implicit `to` is the peers in a topic mesh, although having a specific to is also useful (the Go implementation supports both implicit of all topic peers and explicit peers)

> 

Author: PublishPartial is idempotent. If a peer has sent us a partial iwant, this publish call will respond to it with the relevant parts.

If you call PublishPartial twice with no changes in between, the second time does nothing.

is only an example of one possible API.

Message contents are application defined, thus splitting a message must be
application defined. Applications should provide a Partial Message type that
supports the following operations:

1. `.GroupID() -> GroupID: bytes`
2. `.PartialMessageBytes(partsMetadata: bytes) -> Result<(EncodedPartialMessage: bytes), Error>`
   1. The method should return an encoded partial message with just the parts the
      peer requested.
3. `.PartsMetadata() -> bytes`

Gossipsub in turn provides a `.PublishPartial(PartialMessage, PartialPublishOptions)` method.

The `PartialPublishOptions` contains:

1. Optional eager data that should be pushed to peers who haven't sent us a bitmap yet.
2. An optional list of peers to publish to instead of the topic mesh peers.
    1. This is useful for responding to peers who are not in the node's mesh, but
      sent the node a PartialMessage (e.g similar to Gossipsub's `IHAVE`)

When Gossipsub receives a partial message it MUST forward it to the application.
The application decides if it should act on the message by either requesting
parts or forwarding the message. Both are done with `.PublishPartial`.

> Gossipsub MUST forward all messages to the application, not just messages from

Reviewer 4: is this not just from mesh peers?

Author: yes, thank you.

# Partial Messages Extension

| Lifecycle Stage | Maturity      | Status | Latest Revision |
| --------------- | ------------- | ------ | --------------- |
| 1A              | Working Draft | Active | r0, 2025-06-23  |

> Authors: [@marcopolo, @cskiraly]

Reviewer 3: ```suggestion
Authors: [@marcopolo], [@cskiraly]
```

> Interest Group: TODO

Reviewer 3: ```suggestion
Interest Group: [@jxs], [@dknopik]
```

[@marcopolo]: https://github.com/marcopolo
[@cskiraly]: https://github.com/cskiraly

> 

Reviewer 3: ```suggestion
[@jxs]: https://github.com/jxs
[@dknopik]: https://github.com/dknopik
```

| --------------- | ------------- | ------ | --------------- |
| 1A              | Working Draft | Active | r0, 2025-06-23  |

Authors: [@marcopolo], [@cskiraly]

Interest Group: [@jxs], [@dknopik], [@sukunrt], [@raulk]

[@marcopolo]: https://github.com/marcopolo
[@cskiraly]: https://github.com/cskiraly
[@jxs]: https://github.com/jxs
[@raulk]: https://github.com/raulk
[@dknopik]: https://github.com/dknopik
[@sukunrt]: https://github.com/sukunrt

See the [lifecycle document][lifecycle-spec] for context about the maturity level
and spec status.

[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md

## Overview

Partial Messages Extensions allow users to transmit only a small part of a
message rather than a full message. This is especially useful in cases where
there is a large message and a peer is missing only a small part of the
message.

## Terms and Definitions

**Full Message**: A Gossipsub Message.

**Message Part**: The smallest verifiable part of a message.

**Partial Message**: A group of one or more message parts.

**Group ID**: An identifier to some Full Message. This must not depend on

> knowing the full message, so it can not simply be a hash of the full message.

Reviewer: Can you add Eager Data here?

  defined, partial messages allow cross-forwarding cells between these topics
  [2].

## Advantage of Partial Messages over smaller Gossipsub Messages

Partial Messages within a group imply some structure and correlation. Thus,
multiple partial messages can be referenced succinctly. For example, parts can
be referenced by bitmaps, ranges, or a bloom filter.

The structure of partial messages in a group, as well as how partial messages
are referenced is application defined.

If, in some application, a group only ever contained a single partial message,
then partial messages would be the same as smaller messages.

## Protocol Messages

The following section specifies the semantics of each field in the protocol
message.

### partialMessage

The `partialMessage` field encodes one or more parts of the full message. The
encoding is application defined.

### partsMetadata

The `partsMetadata` field encodes the parts a peer has and wants. The encoding
is application defined. An unset value carries no information besides that the
peer did not send a value.

Upon receiving a `partsMetadata` a node SHOULD respond with only parts the peer

> wants.

Reviewer: ```suggestion
Upon receiving a `partsMetadata` a node SHOULD respond with only parts the peer
doesn't have.
```

A later `partsMetadata` replaces a prior one.

`partsMetadata` can be used during heartbeat gossip to inform non-mesh topic

> peers about parts this node has.

Reviewer 3: ```suggestion
During heartbeat gossip, partsMetadata can be used to inform a random subset of non-mesh topic peers about the parts held by this node, similar to full message IHAVE gossip.
```

Implementations are free to select when to send an update to their peers based
on signaling bandwidth tradeoff considerations.

### Changes to `SubOpts` and interaction with the existing Gossipsub mesh.

The `SubOpts` message is how a peer subscribes to a topic.

Partial Messages uses the same mesh as normal Gossipsub messages. It is a
replacement to "full" messages. A node requests a peer to send partial messages
for a specific topic by setting the `requestsPartial` field in the `SubOpts`
message to true. A node signals support for sending partial messages on a given
topic by setting the `supportsSendingPartial` field in `SubOpts` to true. A node can
support sending partial messages without wanting to receive them.

If a node requests partial messages, it MUST support sending partial messages.

A node uses a peer's `supportsSendingPartial` setting to know if it can send a

> partial message request to a peer. It uses its `requestsPartial` setting to know

Reviewer 5: What does "partial message request" mean? IIUC, it means sending `partsMetadata` to request missing parts from a peer. If so, I'd suggest making this explicit: 

```suggestion
A node uses a peer's `supportsSendingPartial` setting to know if it can send `partsMetadata` to a peer. It uses its `requestsPartial` setting to know
```

Author: Thanks, updated in the latest commit.

whether to send the peer a full message or a partial message.

> If a peer supports partial messages on a topic but did not request them, a node

Reviewer: ```suggestion
If a peer supports sending partial messages on a topic but did not request them, a node
```

| \*                       | The receiver SHOULD send full messages.                                   |

| SubOpts.supportsSendingPartial | behavior of receiver that requested partial messages for the topic                                               |
| ------------------------ | ---------------------------------------------------------------------------------------------------------------- |
| true                     | The receiver expects the peer to respond to partial message requests, and receive `partsMetadata` from the peer. |
| false                    | The receiver expects full messages.                                                                              |

| SubOpts.supportsSendingPartial | behavior of receiver that did not request partial messages for the topic |
| ------------------------ | ------------------------------------------------------------------------ |
| \*                       | The receiver expects full messages                                       |

## Partial Message Gossip

Partial Messages can replace Gossipsub's IHAVE/IWANT with a message that

> provides more context (via the Group ID) and allows for partial responses.

Reviewer: ```suggestion
Partial Messages SHOULD replace Gossipsub's IHAVE/IWANT with a message that
provides more context (via the Group ID) and allows for partial responses.
```

When Gossiping, a node that supports partial messages SHOULD NOT send an `IHAVE`
to a peer that requested partial messages. The node SHOULD send a partial message
instead.

## Implementation Recommendations

The following section is not intended to be normative, it is only meant to
provide rough recommendations to implementations.

### Sending `partsMetadata`

Implementations should send their `partsMetadata` whenever it changes. The goal
being to provide an up-to-date view of its parts to its peers. Implementations
should not send duplicate `partsMetadata` to peers when nothing has changed.

If a node has previously sent a peer its `partsMetadata`, and that peer has
responded with parts, the node can assume the peer's view of itself is the union
of the its previously sent `partsMetadata `and the remote peer's
`partsMetadata`. The node does not need to send this peer its updated
`partsMetadata` if its `partsMetadata` has not changed since the previously sent
`partsMetadata`.

On the other side, when a node responds to a peer's request for parts, it should
update the peer's `partsMetadata` with the union of its `partsMetadata` and the
peers `partsMetadata`.

> 

Reviewer: I think this isn't true anymore. This spec doesn't enforce or even suggest sending all the parts that the node has and the peer doesn't have to the peer. So the union op is incorrect.

Author: you're right. I'll remove this section.

Due to the distributed nature of peers, this might result in a period of time
where a peer's `partsMetadata` is not exactly correct, but it should become
eventually consistent.

### Merging Eager Data with a peer's `partsMetadata`.

When eagerly sending data, a node should track which parts it has sent to the
peer in order to update its view of the peer's `partsMetadata`. This allows the
node to avoid sending duplicate data in the case that a peer concurrently sends
a `partsMetadata` that requests data already eagerly sent.

### Reacting to `IHAVE`

If a node requests partial messages and is connected partial message capable
peers, it MAY prefer to delay reacting to a peer's `IHAVE` message in order to
give the opportunity for a partial message request to finish and provide the

> missing message more efficiently.

Reviewer: ```suggestion
If a node is reconstructing a message with partial message extension, it MAY prefer to delay reacting to a peer's `IHAVE` message in order to
give the opportunity for a partial message request to finish and get the message more efficiently.
```

### DoS Resiliency

- Limit the amount of peer initiated state you track.

### Eager pushing data

If you have a peer's `partsMetadata` you should use that to decide what to send
to the peer. The corollary is that you do not eager push data to a peer that has
given you its `partsMetadata`.

### Example Application Interface

Message contents are application defined, thus splitting a message must be
application defined. Applications should provide a Partial Message type that
supports the following operations:

1. `.GroupID() -> GroupID: bytes`
2. `.PartialMessageBytes(partsMetadata: bytes) -> Result<(EncodedPartialMessage: bytes), Error>`
   1. The method should return an encoded partial message with just the parts the
      peer requested.
3. `.EagerPartialMessageBytes() -> Result<(EncodedPartialMessage: bytes, partsMetadata: bytes), Error>`
   1. The method should return an encoded partial message of eager data that
      should be sent, along with the partsMetadata a peer would have after
      receiving the encoded partial message.

> 4. `.PartsMetadata() -> bytes`: The parts this partial message has.

Reviewer 3: this API is now different in the `rust-libp2p`  [implementation](https://github.com/jxs/rust-libp2p/blob/gossipsub-partial-messages/protocols/gossipsub/src/extensions/partial_messages.rs#L72-L93).

Author: I'll rework this section from our latest learnings.

Gossipsub in turn provides a `.PublishPartial(PartialMessage)` method. This
method should be idempotent. The application should be able to call it multiple
times, and the library should make sure not to send redundant or duplicate
messages to peers.

When Gossipsub receives a partial message it MUST forward it to the application.
The application decides if it should act on the message by either requesting

> parts or forwarding the message. Both are done with `.PublishPartial`.

Reviewer: Should this be outside of the implementation guidelines heading?

Author: This is up to the implementation, right?

Reviewer: this specific part isn't. If the partsMetadata is opaque, the implementation MUST necessarily invoke the application to see what to do with the incoming partial message.

Gossipsub MUST forward all messages to the application, not just messages from

> mesh peers.

Reviewer: Same comment on this one. Should this be outside implementation guidelines?

### Fanout and Gossip messages

> 

Reviewer: Same comment here. Move it outside implementation recommendations.

Author: I'm not sure, fanout and gossip and mesh peers are identical from the protocol's perspective. How this is presented is up to the implementation. What do you think should go outside the implementation guidelines section?

Reviewer: on this point + https://github.com/libp2p/specs/pull/685#discussion_r2789068407 combined:

I would prefer calling out that a partial message irrespective of the sender (gossip, mesh, fanout) MUST be pushed to the application. There's nothing gossipsub layer can do here.

Fanout and Gossip messages by definition come from non-mesh peers. Partial

> messages, without eager data, require an exchange of bitmaps before parts are

Reviewer 3: ```suggestion
messages without eager data, require an exchange of bitmaps before parts are
```

transferred. In order for fanout and gossip messages to be useful, the Gossipsub
implementation SHOULD include the peers that sent them in the list of peers to
publish partial messages to. This allows the application to simply call

> `PublishPartial` to respond to both mesh and non mesh peers.

Reviewer: NIT
```suggestion
Fanout and Gossip messages by definition come from non-mesh peers. After receiving partial messages from Fanout and Gossip peers, the Gossipsub
implementation SHOULD include these non mesh peers that sent them in the list of peers to
publish partial messages to. This allows the application to simply call
`PublishPartial` to respond to both mesh and non mesh peers.
```

## Upgrading a topic to use partial messages

Rolling out partial messages on an existing topic allows for incremental
migration with backwards compatibility. The steps are as follows:

1. Deploy nodes that support partial messages, but do not request them for the
   target topic. The goal is to seed support for partial messages before making
   the switch. Nodes signal their support for partial messages by setting the
   subscribe option `supportsSendingPartial` to true.
2. Slowly deploy and monitor nodes that request (and implicitly support) partial
   messages. These nodes should find peers that send them partial messages from
   the previous step. Nodes request partial messages by setting the subscribe
   option `requestPartial` to true.

### Supporting both full and partial messages for a topic

> Partial messages use the same mesh as "full" messages. Supporting both is

Reviewer 3: should it be
```suggestion
Partial messages use the same mesh as full messages. Supporting both is
```
instead?
as we already define above the concept of full messages

straightforward. If a peer subscribes to a topic with a "requestPartial", the
node SHOULD send the peer partial messages. Otherwise, send the node full
messages.

On the receiving side, if the node is in a mixed network of partial and full
messages, and it requests partial messages, the node MUST support receiving full
messages.

## Creating a topic to only use partial messages

> There is currently no mechanism to requre that a topic only be used for partial

Reviewer 3: ```suggestion
There is currently no mechanism to require that a topic only be used for partial
```

### Changes to `SubOpts` and interaction with the existing Gossipsub mesh.

The `SubOpts` message is how a peer subscribes to a topic.

Partial Messages uses the same mesh as normal Gossipsub messages. It is a
replacement to "full" messages. A node requests a peer to send partial messages
for a specific topic by setting the `requestsPartial` field in the `SubOpts`
message to true. A node signals support for sending partial messages on a given
topic by setting the `supportsPartial` field in `SubOpts` to true. A node can
support sending partial messages without wanting to receive them.

> If a node requests partial messages, it MUST support partial messages.

Reviewer: ```suggestion
If a node requests partial messages, it MUST support sending partial messages.
```
